### PR TITLE
[PM-2289] [PM-2293] TDE Login with device Admin Request

### PR DESCRIPTION
**Pull request by andrebispo5**
Merge After https://github.com/bitwarden/mobile/pull/2623
## Type of change
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective

Change logic on Passwordless login to work with TDE **approve with device** and **request admin approval**.

## Code changes

**- src/Core/Services/AuthService.cs:**
Added different decryption methods for TDE options.

## Before you submit
- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: src/Core/Services/AuthService.cs [hunk 1]
```diff
@@ -198,12 +198,32 @@ private async Task<bool> RequirePasswordChangeAsync(string email, string masterP
...
+                else
+                {
+                    var userKey = await _cryptoService.DecryptUserKeyWithMasterKeyAsync(new MasterKey(decryptedKey));
+                    await _cryptoService.SetUserKeyAsync(userKey);
+                }
+                await _deviceTrustCryptoService.TrustDeviceIfNeededAsync();
+            }
+            else
+            {
+                var decKey = await _cryptoService.RsaDecryptAsync(masterKey, decryptionKey);
```
Comment (jlf0dev) [inline src/Core/Services/AuthService.cs:221]: Do we need to account for the auth request holding just the `UserKey` here too?

**Review by jlf0dev — COMMENTED** [review]
Looking good so far, just in case here's the final clients version that just got merged as a reference: bitwarden/clients#5864

File: src/App/Pages/Accounts/LoginPasswordlessRequestViewModel.cs [hunk 2]
```diff
@@ -255,6 +261,15 @@ private async Task CreatePasswordlessLoginAsync()
...
             {
+                //TODO TDE if is admin type save to memory to later see if it was approved
+                /*
+                  const adminAuthReqStorable = new AdminAuthRequestStorable({
+                      id: reqResponse.id,
+                      privateKey: this.authRequestKeyPair.privateKey,
+                    });
+
+                    await this.stateService.setAdminAuthRequest(adminAuthReqStorable);
+                */
```
Comment (fedemkr) [inline src/App/Pages/Accounts/LoginPasswordlessRequestViewModel.cs:272]: ❓ Will this be addressed in a different PR? If so, could we add a reference to the task id here if such task exists?

File: src/App/Pages/Accounts/LoginPasswordlessRequestViewModel.cs [hunk 3]
```diff
@@ -221,6 +221,13 @@ private async Task CheckLoginRequestStatus()
                 var authResult = await _authService.LogInPasswordlessAsync(Email, _requestAccessCode, _requestId, _requestKeyPair.Item2, response.Key, response.MasterPasswordHash);
                 await AppHelpers.ResetInvalidUnlockAttemptsAsync();
 
+                if(authResult == null && await _stateService.IsAuthenticatedAsync())
```
Comment (fedemkr) [inline src/App/Pages/Accounts/LoginPasswordlessRequestViewModel.cs:224]: ⛏️ Add space after the `if`, I think this should have been caught by the formatter:
```suggestion
                if (authResult == null && await _stateService.IsAuthenticatedAsync())
```

File: src/Core/Services/AuthService.cs [hunk 4]
```diff
@@ -198,12 +198,32 @@ private async Task<bool> RequirePasswordChangeAsync(string email, string masterP
...
 
-        public async Task<AuthResult> LogInPasswordlessAsync(string email, string accessCode, string authRequestId, byte[] decryptionKey, string userKeyCiphered, string localHashedPasswordCiphered)
+        public async Task<AuthResult> LogInPasswordlessAsync(string email, string accessCode, string authRequestId, byte[] decryptionKey, string masterKey, string masterKeyHash)
         {
-            var decKey = await _cryptoService.RsaDecryptAsync(userKeyCiphered, decryptionKey);
-            var decPasswordHash = await _cryptoService.RsaDecryptAsync(localHashedPasswordCiphered, decryptionKey);
-            return await LogInHelperAsync(email, accessCode, Encoding.UTF8.GetString(decPasswordHash), null, null, null, new MasterKey(decKey), null, null,
+            AuthResult response = null;
+            // On SSO flow user is already AuthN
+            if (await this._stateService.IsAuthenticatedAsync())
```
Comment (fedemkr) [inline src/Core/Services/AuthService.cs:205]: ⛏️We're not using `this` unless strictly necessary:
```suggestion
            if (await _stateService.IsAuthenticatedAsync())
```

**Review by fedemkr — CHANGES_REQUESTED** [review]

File: src/App/Pages/Accounts/LoginPasswordlessRequestViewModel.cs (hunk 2 above)
Comment (andrebispo5) [reply src/App/Pages/Accounts/LoginPasswordlessRequestViewModel.cs:272]: it's already addressed in this pr: https://github.com/bitwarden/mobile/pull/2643

File: src/Core/Services/AuthService.cs [hunk 5]
```diff
@@ -198,12 +198,40 @@ private async Task<bool> RequirePasswordChangeAsync(string email, string masterP
             return !await _policyService.EvaluateMasterPassword(strength.Value, masterPassword, _masterPasswordPolicy);
         }
 
-        public async Task<AuthResult> LogInPasswordlessAsync(string email, string accessCode, string authRequestId, byte[] decryptionKey, string userKeyCiphered, string localHashedPasswordCiphered)
+        public async Task<AuthResult> LogInPasswordlessAsync(string email, string accessCode, string authRequestId, byte[] decryptionKey, string masterKey, string masterKeyHash)
```
Comment (jlf0dev) [inline src/Core/Services/AuthService.cs:201]: Can we rename `masterKey` since it won't always be a Master Key? Something like `encryptedAuthRequestKey` would be better. Also putting some comments in this method would really help readability.

File: src/Core/Services/AuthService.cs [hunk 6]
```diff
@@ -218,10 +218,18 @@ public async Task<AuthResult> LogInPasswordlessAsync(string email, string access
             }
             else
             {
-                var decKey = await _cryptoService.RsaDecryptAsync(masterKey, decryptionKey);
-                var decKeyHash = await _cryptoService.RsaDecryptAsync(masterKeyHash, decryptionKey);
-                response = await LogInHelperAsync(email, accessCode, Encoding.UTF8.GetString(decKeyHash), null, null, null, new MasterKey(decKey), null, null,
-                null, null, authRequestId: authRequestId);
+                if (string.IsNullOrEmpty(masterKeyHash) && decryptionKey != null)
+                {
+                    var decryptedKey = await _cryptoService.RsaDecryptAsync(masterKey, decryptionKey);
```
Comment (fedemkr) [inline src/Core/Services/AuthService.cs:223]: ⛏️ This line gets repeated in the `else` as well, so it can be moved before the `if` given that it's used on both flows.

**Review by fedemkr — CHANGES_REQUESTED** [review]

File: src/Core/Services/AuthService.cs [hunk 7]
```diff
@@ -198,12 +198,39 @@ private async Task<bool> RequirePasswordChangeAsync(string email, string masterP
...
+                    await _cryptoService.SetUserKeyAsync(new UserKey(decryptedKey));
+                }
+                else
+                {
+                    var decKeyHash = await _cryptoService.RsaDecryptAsync(masterKeyHash, decryptionKey);
+                    response = await LogInHelperAsync(email, accessCode, Encoding.UTF8.GetString(decKeyHash), null, null, null, new MasterKey(decryptedKey), null, null,
+                    null, null, authRequestId: authRequestId);
+                }
+            }
+            return response;
```
Comment (fedemkr) [inline src/Core/Services/AuthService.cs:233]: ⛏️  Sorry I missed this in the previous review, but no need to have the `response` object be created before, also the `decryptedKey`  can be moved outside the if (now it's being repeated) and finally, we can do the next thing to improve arrow code.
```suggestion
            var decryptedKey = await _cryptoService.RsaDecryptAsync(encryptedAuthRequestKey, decryptionKey);
            
            // On SSO flow user is already AuthN
            if (await _stateService.IsAuthenticatedAsync())
            {
                if (string.IsNullOrEmpty(masterKeyHash))
                {
                    await _cryptoService.SetUserKeyAsync(new UserKey(decryptedKey));
                }
                else
                {
                    var userKey = await _cryptoService.DecryptUserKeyWithMasterKeyAsync(new MasterKey(decryptedKey));
                    await _cryptoService.SetUserKeyAsync(userKey);
                }
                await _deviceTrustCryptoService.TrustDeviceIfNeededAsync();
                return null;
            }

            if (string.IsNullOrEmpty(masterKeyHash) && decryptionKey != null)
            {
                await _cryptoService.SetUserKeyAsync(new UserKey(decryptedKey));
                return null;
            }
            
            var decKeyHash = await _cryptoService.RsaDecryptAsync(masterKeyHash, decryptionKey);
            return await LogInHelperAsync(email, accessCode, Encoding.UTF8.GetString(decKeyHash), null, null, null, new MasterKey(decryptedKey), null, null,
                null, null, authRequestId: authRequestId);
```

Comment (andrebispo5) [reply src/Core/Services/AuthService.cs]: Beautiful!

File: src/App/Pages/Accounts/LoginPasswordlessRequestViewModel.cs [hunk 8]
```diff
@@ -248,22 +265,66 @@ private async Task CheckLoginRequestStatus()
...
+
+            await HandlePasswordlessLoginAsync(response, pendingRequest == null && _authRequestType == AuthRequestType.AdminApproval);
             await _deviceActionService.HideLoadingAsync();
         }
 
+        private async Task HandlePasswordlessLoginAsync(PasswordlessLoginResponse response, bool createPendingAdminRequest)
+        {
+            if (response == null)
+            {
+                throw new ArgumentNullException(AppResources.GenericErrorMessage);
```
Comment (fedemkr) [inline src/App/Pages/Accounts/LoginPasswordlessRequestViewModel.cs:313]: ⛏️ The parameter of the exception should be the name of the null param
```suggestion
                throw new ArgumentNullException(nameof(response));
```

**Review by fedemkr — CHANGES_REQUESTED** [review]

**Review by fedemkr — APPROVED** [review]
